src/models/thread.ts

```diff
@@ -504,17 +504,43 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
         throw new Error("Unsupported function on the thread model");
     }
 
+    public getEventReadUpTo(userId: string, ignoreSynthesized?: boolean): string | null {
+        const isCurrentUser = userId === this.client.getUserId();
+        if (isCurrentUser && this.lastReply()) {
+            // If the last activity in a thread is prior to the first threaded read receipt
+            // sent in the room we want to consider this thread as read.
+            const beforeFirstThreadedReceipt = (this.lastReply()?.getTs() ?? 0) < this.room.oldestThreadedReceiptTs;
+            if (beforeFirstThreadedReceipt) {
+                return this.timeline.at(-1)?.getId() ?? null;
+            }
+        }
+
+        const readUpToId = super.getEventReadUpTo(userId, ignoreSynthesized);
+
+        // Checking whether the unthreaded read receipt for that user is more recent
+        // than the read receipt inside that thread.
+        if (this.lastReply()) {
+            if (!this.room.unthreadedReceipts.has(userId)) {
+                return readUpToId;
+            }
+
+            const unthreadedReceiptTs = this.room.unthreadedReceipts.get(userId)!.ts;
+            for (let i = this.timeline?.length - 1; i >= 0; --i) {
+                const ev = this.timeline[i];
+                if (ev.getId() === readUpToId) return readUpToId;
+                if (ev.getTs() < unthreadedReceiptTs) return ev.getId() ?? readUpToId;
+            }
+        }
+
+        return readUpToId;
+    }
+
     public hasUserReadEvent(userId: string, eventId: string): boolean {
         if (userId === this.client.getUserId()) {
-            const publicReadReceipt = this.getReadReceiptForUserId(userId, false, ReceiptType.Read);
-            const privateReadReceipt = this.getReadReceiptForUserId(userId, false, ReceiptType.ReadPrivate);
-            const hasUnreads = this.room.getThreadUnreadNotificationCount(this.id, NotificationCountType.Total) > 0;
-
-            if (!publicReadReceipt && !privateReadReceipt && !hasUnreads) {
-                // Consider an event read if it's part of a thread that has no
-                // read receipts and has no notifications. It is likely that it is
-                // part of a thread that was created before read receipts for threads
-                // were supported (via MSC3771)
+            const beforeFirstThreadedReceipt = (this.lastReply()?.getTs() ?? 0) < this.room.oldestThreadedReceiptTs;
+            const unthreadedReceiptTs = this.room.unthreadedReceipts.get(userId)?.ts ?? 0;
+            const beforeLastUnthreadedReceipt = (this?.lastReply()?.getTs() ?? 0) < unthreadedReceiptTs;
```

Comment:
Why are we replacing the `getReadReceiptForUserId` accessors with new maps in `this.room`? how do the two differ? why do we need both? Please could we have comments to document this stuff?

Reply:
I'm not sure I follow where you're getting at with the comment?

Reply:
yes sorry, this wasn't very well expressed.

My comment was really saying two things, neither of which relate directly to this bit of code.

 1. It felt like a retrograde step to go from accessors like `getReadReceiptForUserId` to directly accessing the properties on Room. You've now changed the direct access to go via accessors, so that part is now resolved.

 2. I was struggling to understand how the read receipt information we now track in `Room` related to the information we already track in `ReadReceipt` (and, by inheritance, Thread). I still am a bit confused about that, to be honest. Do the `receipts` in `ReadReceipt` only include those related to either a particular thread, or no thread at all?